common/buildcraft/factory/tile/TileAutoWorkbenchBase.java

```diff
@@ -71,9 +75,73 @@ protected void disableBindings() {
         crafting.disableBindings();
     }
 
+    private boolean tryUseMaterials() {
+        List<ItemStack> materialsNeed = new ArrayList<>();
+        for(int i = 0; i < slots; i++) {
+            ItemStack stack = invBlueprint.getStackInSlot(i);
+            if(stack != null) {
+                boolean merged = false;
+                for(ItemStack s : materialsNeed) {
+                    if(StackUtil.canMerge(stack, s)) {
+                        s.stackSize += stack.stackSize;
+                        merged = true;
+                    }
+                }
+                if(!merged) {
+                    materialsNeed.add(stack.copy());
+                }
+            }
+        }
+        List<Pair<Integer, Integer>> slotUsed = new ArrayList<>();
+        for(int i = 0; i < slots; i++) {
+            ItemStack stack = invMaterials.getStackInSlot(i);
+            for(Iterator<ItemStack> iterator = materialsNeed.iterator(); iterator.hasNext(); ) {
+                ItemStack s = iterator.next();
+                if(StackUtil.canMerge(stack, s)) {
+                    int remaining = MathUtils.clamp(stack.stackSize - s.stackSize, 0, 64);
+                    int used = stack.stackSize - remaining;
+                    slotUsed.add(new ImmutablePair<>(i, used));
+                    s.stackSize -= used;
+                    if(s.stackSize == 0) {
+                        iterator.remove();
+                    }
+                }
+            }
+        }
+        if(materialsNeed.isEmpty()) {
+            for(Pair<Integer, Integer> pair : slotUsed) {
+                ItemStack stack = invMaterials.getStackInSlot(pair.getLeft());
+                stack.stackSize -= pair.getRight();
+                invMaterials.setStackInSlot(pair.getLeft(), stack);
+                if(stack.stackSize == 0) {
+                    invMaterials.setStackInSlot(pair.getLeft(), null);
+                }
+            }
+        }
+        return materialsNeed.isEmpty();
+    }
+
     @Override
     public void update() {
         deltaManager.tick();
+        if(worldObj.isRemote) {
+            return;
+        }
+        if(currentRecipe != null) {
+            ItemStack resultStack = invResult.getStackInSlot(0);
+            if(resultStack == null || currentRecipe.getRecipeOutput().stackSize + resultStack.stackSize <= currentRecipe.getRecipeOutput().getMaxStackSize()) {
+                if(tryUseMaterials()) {
+                    if(resultStack != null) {
+                        if(StackUtil.canMerge(resultStack, currentRecipe.getRecipeOutput())) {
+                            resultStack.stackSize += currentRecipe.getRecipeOutput().stackSize;
+                        }
+                    } else {
+                        resultStack = currentRecipe.getRecipeOutput().copy();
+                    }
+                }
+            }
+            invResult.setStackInSlot(0, resultStack);
+        }
```

Comment:
OK, now I use getCraftingResult, but I still not understand, how to use getRemainingItems.